### PR TITLE
Login backend

### PR DESCRIPTION
**Pull request by xleonx0x**
used bcrypt for hashing #️⃣

### CODE REVIEW

**Comment by xleonx0x** [comment]
**note** haven't had time to test it with database data so probably best to hold off merging until that can happen

**Comment by xleonx0x** [comment]
Haven't implemented tokens yet, but feel free to still look at the database lookup for matching pass / user for register and login

File: backend/src/index.ts [hunk 1]
```diff
@@ -1,54 +1,94 @@
...
+app.post("/register", async (req: TypedRequest<LoginBody>, res: Response) => {
+  const { username, password } = req.body;
+
+  const querySnapshot = await getDocs(collection(db, "users"));
+  if (querySnapshot.docs.some((doc) => doc.data().username === username)) {
+    return res.status(400).send("Username already exists");
+  }
+
+  const saltRounds: number = 10;
+  const hashedPassword: string = await bcrypt.hash(password, saltRounds);
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: although bcrypt adds one too, i'd recommend that we add a random [salt](https://en.wikipedia.org/wiki/Salt_(cryptography)) to the password ourselves as well while hashing for extra security. this salt can then be stored in the user obj in the db

File: backend/src/index.ts [hunk 2]
```diff
@@ -1,54 +1,94 @@
...
+  const { username, password } = req.body;
+  const users = collection(db, "users");
+  const loginDetails = query(users, where("username", "==", username));
+  const details = await getDocs(loginDetails);
+
+  if (details.empty) {
+    return res.status(400).send("Username not found");
+  }
+
+  bcrypt.compare(password, details.docs[0].data().password, (err, result) => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: just note that if we add a salt above, we'll also need to add it here to the entered pw

File: backend/src/index.ts [hunk 3]
```diff
@@ -1,54 +1,94 @@
...
+    colour,
   });
-  console.log('Subscribed! Doc written: ', docRef.id);
-  res.status(200).send(docRef)
+  console.log("Subscribed! Doc written: ", docRef.id);
+  res.status(200).send(docRef);
 });
 
-app.get('/colour', async (req, res) => {
+app.get("/colour", async (req, res) => {
```
Comment (lachlanshoesmith) [inline backend/src/index.ts]: u can delete this haha

**Review by lachlanshoesmith — CHANGES_REQUESTED** [review]
this looks good so far!

File: backend/src/index.ts (hunk 1 above)
Comment (xleonx0x) [reply backend/src/index.ts]: so we salt the password + salt it with bcrypt again?

Comment (lachlanshoesmith) [reply backend/src/index.ts]: yes please! this doesnt make a huge difference to our security really - diminishing returns - *but* it's a common practice you'll likely follow when encrypting/hashing with other libraries in the future

**Review by lachlanshoesmith — APPROVED** [review]

**Pull request merged**